staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go

```diff
@@ -0,0 +1,176 @@
+	parentEntries           []metav1.ManagedFieldsEntry
+	parentGroupVersion      schema.GroupVersion
+	subresourceGroupVersion schema.GroupVersion
+
+	parentPath      fieldpath.Path
+	subresourcePath fieldpath.Path
+}
+
+// ScalePathMappings maps a group/version to the replicas path in a scale subresource
+type ScalePathMappings map[schema.GroupVersion]fieldpath.Path
+
+// NewSubresourceScaler creates a new SubresourceScaler
+func NewSubresourceScaler(parentEntries []metav1.ManagedFieldsEntry, parentGroupVersion schema.GroupVersion, parentPath fieldpath.Path) (*SubresourceScaler, error) {
+	h := SubresourceScaler{
+		parentEntries:      parentEntries,
+		parentGroupVersion: parentGroupVersion,
+	}
+
+	// The assumption is that scale objects have autoscaling/v1 group version and
+	// corresponding .spec.replicas path
+	h.subresourceGroupVersion = schema.GroupVersion{Group: "autoscaling", Version: "v1"}
+	h.subresourcePath = fieldpath.MakePathOrDie("spec", "replicas")
+	h.parentPath = parentPath
+
+	return &h, nil
+}
+
+// ToSubresource filter the managed fields of the main resource and convert
+// them so that they can be handled by a subresource.
+// For the managed fields that have a path that could be handled by a
+// subresource this method performs two changes:
+// 1. The APIVersion is changed to the APIVersion of the subresource
+// 2. The Fields are transformed to the path of that subresource
+func (h *SubresourceScaler) ToSubresource() ([]metav1.ManagedFieldsEntry, error) {
+	var updatedEntries []metav1.ManagedFieldsEntry
+
+	for _, entry := range h.parentEntries {
+		if entry.FieldsV1 == nil {
+			continue
+		}
+
+		fieldset, err := internal.FieldsToSet(*entry.FieldsV1)
+		if err != nil {
+			return nil, err
+		}
+
+		if fieldset.Has(h.parentPath) {
+			newEntry := *entry.DeepCopy()
+			fields, err := h.subresourceFields()
+			if err != nil {
+				return nil, err
+			}
+			newEntry.FieldsV1 = fields
+			newEntry.APIVersion = h.subresourceGroupVersion.String()
+			updatedEntries = append(updatedEntries, newEntry)
+		}
+	}
+	return updatedEntries, nil
+}
+
+// ToParent merges the `subresourceEntries` with the parent entries and
+// transform them accordingly
+func (h *SubresourceScaler) ToParent(subresourceEntries []metav1.ManagedFieldsEntry) ([]metav1.ManagedFieldsEntry, error) {
+
+	if !pathInEntries(subresourceEntries, h.subresourcePath) {
+		return h.parentEntries, nil
+	}
+
+	updatedEntries, err := stealPath(h.parentEntries, h.parentPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the entries from subresource to parent and add them to the
+	// returned entries
+	for _, entry := range subresourceEntries {
+		fieldset, err := internal.FieldsToSet(*entry.FieldsV1)
+		if err != nil {
+			return nil, err
+		}
+		if fieldset.Has(h.subresourcePath) {
+			fields, err := h.parentFields()
+			if err != nil {
+				return nil, err
+			}
+			entry.FieldsV1 = fields
+		}
+		entry.APIVersion = h.parentGroupVersion.String()
```

Comment:
This is possibly changing the version that these fields are tracking, which is incorrect.

Reply:
Each managedFields entry can map to a different version of the object.